### PR TITLE
New uniform treatment for misplaced attribute warnings

### PR DESCRIPTION
**Pull request by ccasin**
The current system for issuing misplaced attribute warnings (warning 53) for
builtin attributes is ad-hoc and missing many cases.  Most attributes can never
trigger this warning (even though there are many places they can appear where
they will be ignored).  For the few that can, there are custom functions that
look at the structure of the program and try to determine where the attribute
makes semantic sense.

This PR replaces that system with a more uniform approach: We record all
attributes that appear in the program as they are parsed, and simply note when
the compiler uses them.  At the end of compilation, misplaced attribute warnings
are issued for those attributes that are never used.  This is similar to the
approach taken by ppxlib.

The biggest changes are in `translattribute` and `builtin_attributes`.

- `builtin_attributes`:
  - This has the big table of attributes from the program being compiled that we
    want to make sure are used.  It's a hashtbl rather than a hashset because
    hashset isn't in stdlib - ~the value is a meaningless bool~.  Attributes are
    removed from the table when a use is seen.  The new function `warn_unused`
    is called at the end of compiliation, issuing "misplaced attribute" warnings
    for anything left in the table.

  - The new function `mk_internal` is used by the parser (and in a few other
    places) whenever an attribute is created that we want to make sure is used.

  - There are two new lookup functions - `has_attribute` and
    `filter_attributes` - that are now used everywhere the compiler looks for
    attributes, and handle marking those attributes used if found.  The
    interface for `filter_attributes` feels a bit baroque, but see the mli
    comment for why and translattribute for motivating uses (simpler suggestions
    welcome).  A third similar function, `when_attribute_is`, is used internally
    but not exposed in the mli (just because it isn't needed elsewhere).

  - The `deprecated`, `alert` and `warn_on_literal_pattern` attributes are
    special cases.  The compiler looks for these not when compiling the term
    they appear on, but rather when compiling other things that refer to the
    term they appear on.  E.g., a value may be labelled deprecated, and if that
    value is never referred to elsewhere we'll never use the deprecated
    attribute.  So, without a special case it would never get removed from the
    table and we'd issue a misplaced attribute warning even though it's
    perfectly fine if the deprecated thing isn't used anywhere.  So, the new
    functions `mark_alerts_used` and `mark_warn_on_literal_pattern_used` are
    called when terms that can have these attributes are added to the
    environment, effectively marking these attributes used if they appear on
    things that can be referred to through the env, which seems like the correct
    behavior.

- `translattribute`:
  - Refactored to use `Builtin_attributes` functions to find attributes, rather
    than ad-hoc list traversals and string pattern matching.  This allows us to
    keep the logic for marking attributes used confined to `Builtin_attributes`,
    and is cleaner anyway.

  - As discussed above, all logic for detecting unused/misplaced attributes has
    been deleted - this is now handled by the new mechanism.  There is a still a
    little logic here for detecting duplicated attributes (no new uniform system
    for that in this patch - would be hard).

  - The purpose of this module is to find attributes in the typed tree that are
    relevant to the middle-end, and to translate them from attributes to
    datatypes used by the middle end.  This module used to remove those
    attributes from the typed tree as it went.  I've changed it not to -
    attributes are not generally removed as they are used in other parts of the
    compiler, and not doing it is simpler and more efficient.

  - Some care had to be taken to avoid "eagerly" looking up attributes before
    we're sure we have a term they make sense for (because now looking them up
    marks them used).  This is why, e.g., `add_inline_attribute` was modified to
    move the check for a function before the attribute lookup, rather than doing
    them simultaneously.

- `tests/warnings/w53.ml`:
  - To confirm this was all working, I added tests for every builtin
    attribute showing places where it can appear (and no warning is triggered)
    and places where it cannot appear (and the warning is now triggered).

### CODE REVIEW

File: lambda/translattribute.ml [hunk 1]
```diff
@@ -17,40 +17,38 @@ open Typedtree
...
+  let inline_attribute = Builtin_attributes.filter_attributes p attributes in
   let attr =
     match inline_attribute with
     | [] -> None
     | [attr] -> Some attr
-    | _ :: {Parsetree.attr_name = {txt;loc}; _} :: _ ->
+    | attr :: {Parsetree.attr_name = {txt;loc}; _} :: _ ->
       Location.prerr_warning loc (Warnings.Duplicated_attribute txt);
-      None
+      Some attr
```
Comment (stedolan) [inline lambda/translattribute.ml:49]: Switching from `None` to `Some attr` looks like a bugfix - did this really ignore attributes that were listed twice, or am I reading the old code wrong?

**Review by stedolan — APPROVED** [review]

File: parsing/builtin_attributes.ml [hunk 2]
```diff
@@ -15,6 +15,84 @@
...
+let builtin_attrs =
+  let tbl = Hashtbl.create 128 in
+  List.iter (fun attr -> Hashtbl.add tbl attr true) builtin_attrs;
+  tbl
+
+let is_builtin_attr s = Hashtbl.mem builtin_attrs s
+
+let mk_internal ?(loc= !default_loc) name payload =
+  if is_builtin_attr name.txt
+  then Attribute_table.add unused_attrs name true;
```
[the remark this answers is not in the document]
Comment (ccasin) [reply parsing/builtin_attributes.ml]: Good point, fixed.

File: lambda/translattribute.ml (hunk 1 above)
Comment (stedolan) [reply lambda/translattribute.ml:49]: I think the new behaviour is much better. Arbitrarily picking between two attributes isn't great, but in the old behaviour `[@inline][@inline]` means "don't inline", which is nonsense.

Also, the choice isn't completely arbitrary - I think parse order is preserved, so the attribute that wins is presumably whichever's first in the source code, which seems reasonable. (Or maybe it's last-wins, I haven't checked whether this list is built in reverse order. Last-wins is also fine by me).

File: parsing/builtin_attributes.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (stedolan) [reply parsing/builtin_attributes.ml]: Thanks!

Just had another look at `builtin_attributes.ml`, and I think the same comment applies to `builtin_attrs`

[the remark this answers is not in the document]
Comment (ccasin) [reply parsing/builtin_attributes.ml]: :) OK, fixed.  Let's see if you can find a third hashmap in there.

**Pull request merged**